modules/storages/app/services/storages/upload_file_service.rb

```diff
@@ -0,0 +1,173 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  class UploadFileService < BaseService
+    using Peripherals::ServiceResultRefinements
+
+    def self.call(container:, project_storage:, file_path:, filename:, file_data:)
+      new(project_storage, container).call(file_path:, filename:, file_data:)
+    end
+
+    def initialize(project_storage, container)
+      super()
+      @project_storage = project_storage
+      @storage = project_storage.storage
+      @container = container
+      @user = determine_user(container)
+    end
+
+    def call(file_path:, filename:, file_data:)
+      with_tagged_logger do
+        info "Starting file upload for #{filename} to #{file_path}"
+
+        unless nextcloud_storage?
+          @result.errors.add(:base, :unsupported_storage_type)
+          @result.success = false
+          return @result
+        end
+
+        ensure_validity_period_exists
+
+        folder = fetch_or_create_folder(file_path).value_or { return @result }
+        file = upload_file(folder, filename, file_data).value_or { return @result }
+        file_link = create_file_link(file).value_or { return @result }
+
+        @result.success = file_link
+        @result
+      end
+    end
+
+    private
+
+    def auth_strategy = Adapters::Registry["nextcloud.authentication.userless"].call
+
+    def nextcloud_storage?
+      @storage.short_provider_type == "nextcloud"
```

Comment:
:yellow_circle: This code already exists

```suggestion
      @storage.provider_type_nextcloud?
```